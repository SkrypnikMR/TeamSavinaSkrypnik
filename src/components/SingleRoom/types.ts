import React from 'react';

export type TSingleRoom = {
    creatorLogin: string;
    gameType: string;
    userLogin: string;
    id: string;
    playWithBot: (arg: string) => void,
    joinRoom: (arg: string) => void
}

export type TDistResult = {
    content: string;
    onClickFunc: (e: React.MouseEvent<Element, MouseEvent>) => void
}
