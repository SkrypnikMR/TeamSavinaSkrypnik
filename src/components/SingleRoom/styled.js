import styled from 'styled-components';

import { midGrey } from '../UI/baseLayout';

export const StSingleRoom = styled.div`
    background: ${midGrey};
    margin: 5px;
    padding: 10px;
    border-radius: 5px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    cursor: pointer;
    :hover {
        -webkit-transform: scale(1.005);
        -ms-transform: scale(1.005);
        transform: scale(1.005);
        box-shadow: 0 14px 28px rgba(0, 0, 0, 0.6), 0 10px 10px rgba(0, 0, 0, 0.22);
        transition-duration: 0.3s;
    }
`;
