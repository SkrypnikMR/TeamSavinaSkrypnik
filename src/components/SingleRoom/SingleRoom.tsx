import React from 'react';
import { useTranslation } from 'react-i18next';
import Button from '../UI/Button';
import { StSingleRoom } from './styled';
import { TSingleRoom, TDistResult } from './types';

const SingleRoom = ({
    creatorLogin,
    gameType,
    userLogin,
    id,
    joinRoom,
    playWithBot,
    subscribeRoom,
}: TSingleRoom) => {
    const { t } = useTranslation();
    const roomOwner: string = creatorLogin === userLogin
        ? t('your_room')
        : creatorLogin;
    const onClickDistributor = (
        creatorLogin: string,
        userLogin: string): TDistResult => {
        return creatorLogin === userLogin
            ? {
                content: 'play_with_bot',
                onClickFunc: (e) => playWithBot(e.target.id),
            }
            : {
                content: 'join',
                onClickFunc: (e) => {
                    joinRoom(e.target.id);
                    subscribeRoom(e.target.id);
                },
            };
    };
    const {
        content,
        onClickFunc } = onClickDistributor(creatorLogin, userLogin);

    return (
        <StSingleRoom>
            <p>{roomOwner}</p>
            <p>{t(gameType)}</p>
            <Button content={content} width="150px" bgColorShadow="0 14px 28px rgba(0, 0, 0, 0.6), 0 10px 10px rgba(0, 0, 0, 0.22)" onClick={onClickFunc} id={id}/>
        </StSingleRoom>
    );
};

export default SingleRoom;
