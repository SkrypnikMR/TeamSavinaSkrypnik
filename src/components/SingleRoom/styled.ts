import styled from 'styled-components';

import { SINGLE_ROOM_BG } from '../UI/baseLayout';

export const StSingleRoom = styled.div`
    background: ${SINGLE_ROOM_BG};
    margin: 5px;
    padding: 10px;
    border-radius: 5px;
    display: grid;
    grid-template: 1fr/ 3fr 3fr 1fr;
    align-self: center;
    cursor: pointer;
    :hover {
        -webkit-transform: scale(1.005);
        -ms-transform: scale(1.005);
        transform: scale(1.005);
        box-shadow: 0 14px 28px rgba(0, 0, 0, 0.6), 0 10px 10px rgba(0, 0, 0, 0.22);
        transition-duration: 0.3s;
    }
    p{
        word-wrap: break-word;
    }
`;
