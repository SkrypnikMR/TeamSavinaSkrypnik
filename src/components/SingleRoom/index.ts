import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import SingleRoom from './SingleRoom';
import { joinRoom, playWithBot } from '../../store/game/actions';

const mapDispatchToProps = (dispatch:Dispatch) => ({
    joinRoom: (payload: string) => dispatch(joinRoom(payload)),
    playWithBot: (payload: string) => dispatch(playWithBot(payload)),
});

export default connect(null, mapDispatchToProps)(SingleRoom);
