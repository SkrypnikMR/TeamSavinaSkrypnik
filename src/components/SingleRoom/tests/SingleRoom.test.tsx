import React from 'react';
import SingleRoom from '../SingleRoom';
import { shallowSmart, mountSmart } from '../../../../__tests__/testHelper';


describe('SingleRoom', () => {
    let props;
    beforeEach(() => {
    props = {
        creatorLogin: 'KekShrek',
        gameType: 'Checkers',
        userLogin: 'KekShrek',
        id: '21ue321h3hjk21bh21-asdshjdsad',
        joinRoom: jest.fn(),
        playWithBot: jest.fn(),
    };
        });
    it('Should match snapshot', () => {
        const component = shallowSmart(<SingleRoom {...props}/>);
        expect(component.html()).toMatchSnapshot();
    });
    it('should render StSingleRoom', () => {
        const component = mountSmart(<SingleRoom {...props}/>);
        expect(component.find('styled__StSingleRoom')).toHaveLength(1);
    });
    it('should render Button', () => {
        const component = mountSmart(<SingleRoom {...props}/>);
        expect(component.find('Button')).toHaveLength(1);
    });
    it('should render p with roomOwner t("your_room") ', () => {
        const component = mountSmart(<SingleRoom {...props}/>);
        expect(component.find('p').first().text()).toBe('your_room');
    });
    it('should render p with roomOwner like creatorLogin ', () => {
        props.userLogin = 'abba';
        const component = mountSmart(<SingleRoom {...props}/>);
        expect(component.find('p').first().text()).toBe(props.creatorLogin);
    });
    it('should click on button and call playWithBot func', () => {
        const component = mountSmart(<SingleRoom {...props} />);
        const testEvent = { target: { id: '1' } };
        component.find('Button').props().onClick(testEvent);
        expect(props.playWithBot).toHaveBeenCalledWith(testEvent.target.id);
    });
    it('should click on button and call joinRoom func', () => {
        props.userLogin = 'abba';
        const component = mountSmart(<SingleRoom {...props} />);
        const testEvent = { target: { id: '1' } };
        component.find('Button').props().onClick(testEvent);
        expect(props.joinRoom).toHaveBeenCalledWith(testEvent.target.id);
    });
});
