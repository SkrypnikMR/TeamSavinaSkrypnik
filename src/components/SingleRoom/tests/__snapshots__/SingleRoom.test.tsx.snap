// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleRoom Should match snapshot 1`] = `"<div class=\\"styled__StSingleRoom-sc-1b2p865-0 bJLfHt\\"><p>your_room</p><p>Checkers</p><button id=\\"21ue321h3hjk21bh21-asdshjdsad\\" type=\\"button\\" width=\\"150px\\" class=\\"styled__StButton-sc-1y35zax-0 iuauOw\\">play_with_bot</button></div>"`;
